### PR TITLE
langchain[patch]: Issue #2756 Add Qdrant custom payload on documents to query them by filter

### PR DESCRIPTION
**Pull request by youngjaeheo2002**
Description:
- Added custom JSON payload integration to Qdrant integration 
- This allows querying database with more criteria such as geological distance
- Implemented additional functionality to addDocuments and addVectors functions in langchain/src/vectorstores/qdrant.ts to allow storing vectors with documents that have custom payloads
- Testing:
    - Added unit tests for Qdrant Vector Store adding custom payloads of size 1, size greater than 1, and no custom payload 

- Issue: #2756

### CODE REVIEW

File: langchain/src/vectorstores/qdrant.ts [hunk 1]
```diff
@@ -115,6 +125,8 @@ export class QdrantVectorStore extends VectorStore {
       payload: {
         content: documents[idx].pageContent,
         metadata: documents[idx].metadata,
+        ...(documents[idx].customPayload ? documents[idx].customPayload != customPayload ? 
+          documents[idx].customPayload : customPayload : customPayload),
```
Comment (bracesproul) [inline langchain/src/vectorstores/qdrant.ts:129]: Could you just do something like this?
```suggestion
          documents[idx].customPayload ?? customPayload),
```
Not sure if this is valid though

File: langchain/src/vectorstores/qdrant.ts [hunk 2]
```diff
@@ -204,20 +216,23 @@ export class QdrantVectorStore extends VectorStore {
...
+   * @param customPayload Optional 'object' in JSON format used to query database on more criteria
    * @returns Promise that resolves with a new `QdrantVectorStore` instance.
    */
   static async fromTexts(
     texts: string[],
     metadatas: object[] | object,
     embeddings: Embeddings,
-    dbConfig: QdrantLibArgs
+    dbConfig: QdrantLibArgs,
+    customPayload? : object
```
Comment (bracesproul) [inline langchain/src/vectorstores/qdrant.ts:227]:
```suggestion
    customPayload?: object
```
nit

File: langchain/src/vectorstores/qdrant.ts [hunk 3]
```diff
@@ -204,20 +216,23 @@ export class QdrantVectorStore extends VectorStore {
...
   ): Promise<QdrantVectorStore> {
     const docs = [];
     for (let i = 0; i < texts.length; i += 1) {
       const metadata = Array.isArray(metadatas) ? metadatas[i] : metadatas;
       const newDoc = new Document({
         pageContent: texts[i],
         metadata,
+        customPayload: customPayload,
```
Comment (bracesproul) [inline langchain/src/vectorstores/qdrant.ts:235]:
```suggestion
        customPayload,
```
nit

**Review by bracesproul — COMMENTED** [review]
overall looks good to me once my comments have been resolved
cc @jacoblee93 for a second look

**Review by bracesproul — COMMENTED** [review]
Also one merge conflict, you'll have to push up the update to the `document` file inside `langchain-core` since the contents of that file have been moved.

**Comment by Dominik-luszcz** [comment]
Hi, I added the optional parameter customPayload to the Documents file in langchain-core, but the only test cases that fail are the:
```
expect(docs[0]).toMatchInlineSnapshot(`
    Document {
      "metadata": {
        "blobType": "application/jsonl+json",
        "line": 1,
        "source": "blob",
      },
      "pageContent": "This is a sentence.",
    }
  `);
```
cases in the blob.test.ts files. To pass these test cases, all we need to do is add `"customPayload": undefined` before the "metadata" in the snapshot above. Am I allowed to add this to was the test cases? Or should I try to prevent the undefined values in the loader.load() function?

**Comment by jacoblee93** [comment]
We can't change the core document schema - it seems like it's some custom querying logic? What's the use-case for persisting it like this vs. some metadata field or just passing it in on ingestion?

**Comment by Dominik-luszcz** [comment]
We saw that in the request, it was suggested to have the custom query parameters within the document indicating that it was needed. Under further consideration it seems like this was unnecessary. We can change it so that it does not have it inside the document.

**Comment by Dominik-luszcz** [comment]
We have updated the code. Please give us any feedback when you have the chance.

File: langchain/src/vectorstores/qdrant.ts [hunk 4]
```diff
@@ -84,13 +84,18 @@ export class QdrantVectorStore extends VectorStore {
    * from the documents using the `Embeddings` instance and then adds the
    * vectors to the database.
    * @param documents Array of `Document` instances to be added to the Qdrant database.
+   * @param customPayload Optional 'object' in JSON format used to query database on more criteria
    * @returns Promise that resolves when the documents have been added to the database.
    */
-  async addDocuments(documents: Document[]): Promise<void> {
+  async addDocuments(
+    documents: Document[],
+    customPayload?: object[]
```
Comment (jacoblee93) [inline langchain/src/vectorstores/qdrant.ts:92]: Can you change this to an object with a field called `customPayload`? That better matches the base interface:

https://github.com/langchain-ai/langchainjs/blob/main/langchain-core/src/vectorstores.ts#L126

File: langchain/src/vectorstores/qdrant.ts [hunk 5]
```diff
@@ -204,13 +215,15 @@ export class QdrantVectorStore extends VectorStore {
...
+   * @param customPayload Optional 'object' in JSON format used to query database on more criteria
    * @returns Promise that resolves with a new `QdrantVectorStore` instance.
    */
   static async fromTexts(
     texts: string[],
     metadatas: object[] | object,
     embeddings: Embeddings,
-    dbConfig: QdrantLibArgs
+    dbConfig: QdrantLibArgs,
+    customPayload?: object[]
```
Comment (jacoblee93) [inline langchain/src/vectorstores/qdrant.ts:226]: Would prefer this field to be within `dbConfig`

File: langchain/src/vectorstores/qdrant.ts [hunk 6]
```diff
@@ -230,15 +248,17 @@ export class QdrantVectorStore extends VectorStore {
...
    * @param dbConfig `QdrantLibArgs` instance specifying the configuration for the Qdrant database.
+   * @param customPayload Optional 'object' in JSON format used to query database on more criteria
    * @returns Promise that resolves with a new `QdrantVectorStore` instance.
    */
   static async fromDocuments(
     docs: Document[],
     embeddings: Embeddings,
-    dbConfig: QdrantLibArgs
+    dbConfig: QdrantLibArgs,
+    customPayload?: object[]
```
Comment (jacoblee93) [inline langchain/src/vectorstores/qdrant.ts:258]: See above

**Review by jacoblee93 — COMMENTED** [review]
Thank you for the changes! See comments

**Comment by Dominik-luszcz** [comment]
We have made the changes, please let me know if there is anything wrong with them.

File: langchain/src/vectorstores/qdrant.ts [hunk 7]
```diff
@@ -2,6 +2,7 @@ import { QdrantClient } from "@qdrant/js-client-rest";
 import type { Schemas as QdrantSchemas } from "@qdrant/js-client-rest";
 import { v4 as uuid } from "uuid";
 
+import { AddDocumentOptions } from "closevector-common";
```
Comment (jacoblee93) [inline langchain/src/vectorstores/qdrant.ts:5]: This looks like a mistake?

File: langchain/src/vectorstores/tests/qdrant.test.ts [hunk 8]
```diff
@@ -1,5 +1,6 @@
 /* eslint-disable @typescript-eslint/no-explicit-any */
 import { jest, test, expect } from "@jest/globals";
+import { AddDocumentOptions } from "closevector-common";
```
Comment (jacoblee93) [inline langchain/src/vectorstores/tests/qdrant.test.ts:3]: This looks like a mistake?

**Comment by Dominik-luszcz** [comment]
> Just specify your own add document options type

Something like this inside qdrant.ts or would something else be better?

`export type qdrantDocumentOptions = {
  customPayload: object[];
};`

**Comment by jacoblee93** [comment]
@Dominik-luszcz and @youngjaeheo2002 LMK if this looks good!

**Comment by Dominik-luszcz** [comment]
Yeah it looks good to us! Thanks @jacoblee93!

**Comment by jacoblee93** [comment]
Thank you!

**Pull request merged**